contrib/win32/win32compat/ssh-agent/keyagent-request.c

```diff
@@ -367,6 +541,173 @@ process_remove_all(struct sshbuf* request, struct sshbuf* response, struct agent
+			debug("failed to add key to store");
+			goto done;
+		}
+	}
+
+	debug("added smartcard keys to store");
+
+	memset(&sa, 0, sizeof(SECURITY_ATTRIBUTES));
+	sa.nLength = sizeof(sa);
+	if ((!ConvertStringSecurityDescriptorToSecurityDescriptorW(REG_KEY_SDDL, SDDL_REVISION_1, &sa.lpSecurityDescriptor, &sa.nLength)) ||
+		convert_blob(con, pin, (DWORD)pin_len, &epin, (DWORD*)&epin_len, 1) != 0 ||
+		get_user_root(con, &user_root) != 0 ||
+		RegCreateKeyExW(user_root, SSH_PKCS11_PROVIDERS_ROOT, 0, 0, 0, KEY_WRITE | KEY_WOW64_64KEY, &sa, &reg, NULL) != 0 ||
+		RegCreateKeyExA(reg, canonical_provider, 0, 0, 0, KEY_WRITE | KEY_WOW64_64KEY, &sa, &sub, NULL) != 0 ||
+		RegSetValueExW(sub, L"provider", 0, REG_BINARY, canonical_provider, (DWORD)strlen(canonical_provider)) != 0 ||
+		RegSetValueExW(sub, L"pin", 0, REG_BINARY, epin, (DWORD)epin_len) != 0) {
+		debug("failed to add pkcs11 provider to store");
+		goto done;
+	}
+
+	debug("added pkcs11 provider to store");
+	success = 1;
+done:
+	r = 0;
+	if (request_invalid)
+		r = -1;
+	else if (sshbuf_put_u8(response, success ? SSH_AGENT_SUCCESS : SSH_AGENT_FAILURE) != 0)
+		r = -1;
+
+	/* delete created reg keys if not succeeded*/
+	if ((success == 0) && reg) {
+		if (thumbprint)
+			RegDeleteKeyExA(reg, thumbprint, KEY_WOW64_64KEY, 0);
+		if (canonical_provider)
+			RegDeleteKeyExA(reg, canonical_provider, KEY_WOW64_64KEY, 0);
+	}
+	pkcs11_terminate();
+
+	if (sa.lpSecurityDescriptor)
+		LocalFree(sa.lpSecurityDescriptor);
+	for (i = 0; i < count; i++)
+		sshkey_free(keys[i]);
+	if (thumbprint)
+		free(thumbprint);
+	if (pubkey_blob)
+		free(pubkey_blob);
+	if (provider)
+		free(provider);
+	if (pin)
+		free(pin);
+	if (epin)
+		free(epin);
+	if (user_root)
+		RegCloseKey(user_root);
+	if (reg)
+		RegCloseKey(reg);
+	if (sub)
+		RegCloseKey(sub);
+	return r;
+}
+
+int process_remove_smartcard_key(struct sshbuf* request, struct sshbuf* response, struct agent_connection* con)
+{
+	char *provider = NULL, *pin = NULL, canonical_provider[PATH_MAX];
+	int r = 0, request_invalid = 0, success = 0, index = 0;
+	HKEY user_root = 0;
+
+	if ((r = sshbuf_get_cstring(request, &provider, NULL)) != 0 ||
+		(r = sshbuf_get_cstring(request, &pin, NULL)) != 0) {
+		debug("remove smartcard request is invalid");
+		request_invalid = 1;
+		goto done;
+	}
+
+	if (realpath(provider, canonical_provider) == NULL) {
+		debug("failed PKCS#11 add of \"%.100s\": realpath: %s",
+			provider, strerror(errno));
+		request_invalid = 1;
+		goto done;
+	}
+
+	// Remove 'drive root' if exists
+	if (canonical_provider[0] == '/')
+		memmove(canonical_provider, canonical_provider + 1, strlen(canonical_provider));
+
+	if (get_user_root(con, &user_root) != 0) {
+		success = 1;
+		goto done;
+	}
+
+	if (remove_matching_subkeys_from_registry(user_root, SSH_KEYS_ROOT, L"comment", canonical_provider) != 0 ||
+		remove_matching_subkeys_from_registry(user_root, SSH_PKCS11_PROVIDERS_ROOT, L"provider", canonical_provider) != 0) {
+		goto done;
+	}
```

Comment:
Am I misunderstanding something, or do we leave `pub`, `type`, and `pin` in the registry? Would it make sense to clear them as well?

Reply:
`remove_matching_subkeys_from_registry(root, key, name_value, data_value)` will remove all sub keys in `root` tree under `key` which have both name value equals to `name_value` and data value equals to `data_value`.
The emphasis is that it will remove the entire matching sub key, and all of its (name,data) pairs, which is how `pub`, `type`, and `pin` will also get removed. I've thoroughly tested this.
I've added a couple of `regedit` screenshots to visualize the structure (one provider with one key for the simplicity of the example)
![regedit-openssh](https://user-images.githubusercontent.com/1812956/144995361-3357826b-594c-44fe-ab03-af95b393f9f2.png)

Reply:
Thank you!